examples/dpo_demo_gemma3.ipynb

```diff
@@ -0,0 +1,849 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "reasoning_start = \"<reasoning>\"\n",
+    "reasoning_end = \"</reasoning>\"\n",
+    "solution_start = \"<answer>\"\n",
+    "solution_end = \"</answer>\"\n",
+    "\n",
+    "\n",
+    "SYSTEM_PROMPT = f\"\"\"You are given a problem. Think about the problem and \\\n",
+    "provide your reasoning. Place it between {reasoning_start} and \\\n",
+    "{reasoning_end}. Then, provide the final answer (i.e., just one numerical \\\n",
+    "value) between {solution_start} and {solution_end}.\"\"\"\n",
+    "\n",
+    "TEMPLATE = \"\"\"<start_of_turn>user\n",
+    "{system_prompt}\n",
+    "\n",
+    "{question}<end_of_turn>\n",
+    "<start_of_turn>model\"\"\"\n",
+    "\n",
+    "def generate(\n",
+    "    question, sampler, temperature=0.7, top_k=50, top_p=0.95, seed=None\n",
+    "):\n",
+    "  \"\"\"Given prompt, generates text.\"\"\"\n",
+    "\n",
+    "  if isinstance(question, str):\n",
+    "    input_batch = [\n",
+    "        TEMPLATE.format(\n",
+    "            system_prompt=SYSTEM_PROMPT,\n",
+    "            question=question,\n",
+    "        ),\n",
+    "    ]\n",
+    "  else:\n",
+    "    input_batch = [\n",
+    "        TEMPLATE.format(\n",
+    "            system_prompt=SYSTEM_PROMPT,\n",
+    "            question=q,\n",
+    "        )\n",
+    "        for q in question\n",
+    "    ]\n",
+    "\n",
+    "  out_data = sampler(\n",
+    "      input_strings=input_batch,\n",
+    "      max_generation_steps=TOTAL_GENERATION_STEPS, # Was 768\n",
+    "      temperature=temperature,\n",
+    "      top_k=top_k,\n",
+    "      top_p=top_p,\n",
+    "      echo=False,\n",
+    "      seed=seed if seed is not None else None,\n",
+    "  )\n",
+    "\n",
+    "  output = out_data.text\n",
+    "  if isinstance(question, str):\n",
+    "    return output[0]\n",
+    "  return output"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "match_format = re.compile(\n",
+    "    rf\"^[\\s]{{0,}}\"\n",
+    "    rf\"{reasoning_start}.+?{reasoning_end}.*?\"\n",
+    "    rf\"{solution_start}(.+?){solution_end}\"\n",
+    "    rf\"[\\s]{{0,}}$\",\n",
+    "    flags=re.MULTILINE | re.DOTALL,\n",
+    ")\n",
+    "\n",
+    "match_format.search(\n",
+    "    f\"{reasoning_start}Let me\"\n",
+    "    f\" think!{reasoning_end}{solution_start}2{solution_end}\",\n",
+    ")\n",
+    "\n",
+    "match_numbers = re.compile(\n",
+    "    rf\"{solution_start}.*?([\\d\\.]{{1,}})\", flags=re.MULTILINE | re.DOTALL\n",
+    ")\n",
+    "match_numbers.findall(f\"{solution_start}  0.34  {solution_end}\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def evaluate(\n",
```

Comment:
Do you want to talk about what accuracy we get here? Roughly. We can mention it is around xy.z% or something